src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -40,67 +125,77 @@ func (sm *StorageMinerActor_I) verifyPoStSubmission(postSubmission poster.PoStSu
+	default:
+		// TODO: proper throw
+		panic("invalid state in activateSector")
+	}
+
+	sm.Sectors()[sectorNo].Impl().State_ = SectorActive()
+	sm.SectorTable().Impl().ActiveSectors_ += 1
+
+	// TODO: decide if commit state change here
 }
 
-// failSector moves Sectors from Active/Committed/Recovering into Failing State
-// and increments FaultCount
-func (sm *StorageMinerActor_I) failSectors(sectors []sector.SectorNumber) {
-	// TODO: detemine how much pledge collateral to slash
-	// TODO: SendMessage(spa.SlashPledgeCollateral)
-	var powerUpdate uint64 = 0
+// failSector moves Sector from Active/Committed/Recovering into Failing State
+// and increments FaultCount if asked to do so (DeclareFaults does not increment faultCount)
+// move Sector from Failing to Cleared State if increment results in faultCount exceeds MaxFaultCount
+// update SectorTable
+// remove from ProvingSet
+func (sm *StorageMinerActor_I) failSector(sectorNo sector.SectorNumber, increment bool) {
+	newFaultCount := sm.Sectors()[sectorNo].State().FaultCount
 
-	// this can happen from all states
-	for _, sectorNo := range sectors {
-		state := sm.SectorStates()[sectorNo]
-		switch state.StateNumber {
-		case SectorActiveSN:
-			// SlashStorageDealCollateral
-			// SendMessage(sma.slashStorageDealCollateral(sc.DealIDs()))
-			sm.incrementFaultCount(sectorNo)
-			powerUpdate = powerUpdate - sm.Info().SectorSize()
-		case SectorCommittedSN, SectorRecoveringSN:
-			// SlashStorageDealCollateral
-			// SendMessage(sma.slashStorageDealCollateral(sc.DealIDs()))
-			sm.incrementFaultCount(sectorNo)
-		case SectorFailingSN:
-			sm.incrementFaultCount(sectorNo)
-		default:
-			// TODO: proper failure
-			panic("Invalid sector state in CronAction")
-		}
+	if increment {
+		newFaultCount += 1
 	}
 
-	// TODO
-	// Reset Proving Period and report power updates
-	// sm.ProvingPeriodEnd_ = PROVING_PERIOD_TIME
-	// SendMessage(sma.UpdatePower(powerUpdate))
-}
+	state := sm.Sectors()[sectorNo].State()
+	switch state.StateNumber {
+	case SectorActiveSN:
+		// wont be terminated from Active
+		sm.SectorTable().Impl().ActiveSectors_ -= 1
+	case SectorCommittedSN:
+		sm.SectorTable().Impl().CommittedSectors_ -= 1
+	case SectorRecoveringSN:
+		sm.SectorTable().Impl().RecoveringSectors_ -= 1
+	case SectorFailingSN:
+
+	default:
+		// TODO: proper failure
+		panic("Invalid sector state in CronAction")
+	}
+
+	sm.SectorTable().Impl().FailingSectors_ += 1
 
-// increment FaultCount and check if greater than MaxFaultCount
-// move from Failing to Cleared State if so
-func (sm *StorageMinerActor_I) incrementFaultCount(sectorNo sector.SectorNumber) {
-	newFaultCount := sm.SectorStates()[sectorNo].FaultCount + 1
 	if newFaultCount > MAX_CONSECUTIVE_FAULTS {
 		// TODO: heavy penalization: slash pledge collateral and delete sector
 		// TODO: SendMessage(SPA.SlashPledgeCollateral)
+		sm.SectorTable().Impl().TerminatedSectors_ += 1
+		sm.SectorTable().Impl().FailingSectors_ -= 1
```

Comment:
why not call this within `clearSector` in a state switch? keeps it in the same place. `sm.clearSector` is the state transition. same applies to all similar spots that change counts right before calling `sm.clearSector`